File: 프로그래머스/lv1/42889. 실패율/실패율.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

bool compare(pair<int, double>& a, pair<int, double>& b) {
    if (a.second == b.second) {
        return a.first < b.first;
    }
    return a.second > b.second;
}

vector<int> solution(int N, vector<int> stages) {
    vector<int> lived(N + 1, 0);
    for (int stage : stages) {
        lived[stage - 1]++;
    }
    
    vector<int> success(N + 1, 0);
    success[N] = lived[N];
    for (int i = N; i > 0; i--) {
        success[i - 1] = success[i] + lived[i - 1];
    }
    
    vector<pair<int, double>> failed;
    for (int i = 0; i < N; i++) {
        if (success[i] == 0) {
            failed.push_back({i + 1, 0.0});
        } else {
            failed.push_back({i + 1, static_cast<double>(lived[i]) / success[i]});
        }
    }
    
    sort(failed.begin(), failed.end(), compare);
    
    vector<int> answer;
    for (pair<int, double>& p : failed) {
        answer.push_back(p.first);
    }
    
    return answer;
}

File: 프로그래머스/lv1/42889. 실패율/실패율.py
def solution(N, stages):
    lived = [0]*(N+1)
    for stage in stages:
        lived[stage-1] += 1
    success = lived[:]
    print(lived)

    for i in range(N,0, -1):
        success[i-1] += success[i]
    failed = [[x,0] for x in range(1,N+1)]
    for i in range(N):
        if success[i] == 0:
            continue
        failed[i][1] = lived[i] / success[i]
    failed.sort(key =lambda x : [-x[1], x[0]])
    answer = [x[0] for x in failed]
    return answer